optuna/integration/botorch.py

```diff
@@ -0,0 +1,467 @@
+
+
+@experimental("2.4.0")
+def qehvi_candidates_func(
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+) -> "torch.Tensor":
+    """Quasi MC-based batch Expected Hypervolume Improvement (qEHVI).
+
+    The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
+    with multi-objective optimization.
+
+    .. seealso::
+        :func:`~optuna.integration.botorch.qei_candidates_func` for argument and return value
+        descriptions.
+    """
+
+    n_outcomes = train_obj.size(-1)
+
+    if train_con is not None:
+        train_y = torch.cat([train_obj, train_con], dim=-1)
+
+        is_feas = (train_con <= 0).all(dim=-1)
+        partitioning_y = train_obj[is_feas]
+
+        constraints = []
+        n_contraints = train_con.size(1)
+
+        for i in range(n_contraints):
+            constraints.append(lambda Z, i=i: Z[..., -n_contraints + i])
+        additional_qehvi_kwargs = {
+            "objective": IdentityMCMultiOutputObjective(outcomes=list(range(n_outcomes))),
+            "constraints": constraints,
+        }
+    else:
+        train_y = train_obj
+
+        partitioning_y = train_obj
+
+        additional_qehvi_kwargs = {}
+
+    train_x = normalize(train_x, bounds=bounds)
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.shape[-1]))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_model(mll)
+
+    partitioning = NondominatedPartitioning(num_outcomes=n_outcomes, Y=partitioning_y)
+
+    ref_point = train_obj.min(dim=0).values - 1e-8
+    ref_point_list = ref_point.tolist()
+
+    acqf = qExpectedHypervolumeImprovement(
+        model=model,
+        ref_point=ref_point_list,
+        partitioning=partitioning,
+        sampler=SobolQMCNormalSampler(num_samples=256),
+        **additional_qehvi_kwargs,
+    )
+
+    standard_bounds = torch.zeros_like(bounds)
+    standard_bounds[1] = 1
+
+    candidates, _ = optimize_acqf(
+        acq_function=acqf,
+        bounds=standard_bounds,
+        q=1,
+        num_restarts=20,
+        raw_samples=1024,
+        options={"batch_limit": 5, "maxiter": 200, "nonnegative": True},
+        sequential=True,
+    )
+
+    candidates = unnormalize(candidates.detach(), bounds=bounds)
+
+    return candidates
+
+
+def _get_default_candidates_func(
+    n_objectives: int,
+) -> Callable[
+    [
+        "torch.Tensor",
+        "torch.Tensor",
+        Optional["torch.Tensor"],
+        "torch.Tensor",
+    ],
+    "torch.Tensor",
+]:
+    if n_objectives == 1:
+        return qei_candidates_func
+    elif n_objectives > 1:
+        return qehvi_candidates_func
+    else:
+        assert False, "Should not reach."
+
+
+# TODO(hvy): Allow utilizing GPUs via some parameter, not having to rewrite the callback
+# functions.
+@experimental("2.4.0")
+class BoTorchSampler(BaseMultiObjectiveSampler):
```

Comment:
Cannot be this sampler use with a single `Study` and `FrozenTrial`? It seems that this sampler is only available with the multi-objective module, but we will integrate the multi-objective module with Optuna itself, so this is not a problem after the integration.

Reply:
Yes, we could wait with merging this PR until after https://github.com/optuna/optuna/pull/1994. Then, I could update this sampler to a standard sampler.